### PR TITLE
995 laos eoc weather entities

### PR DESCRIPTION
**Pull request by IgorNadj**
### Changes:

- Add entity type 'city'
- Import 10 most populous cities of Laos

### CODE REVIEW

**Comment by IgorNadj** [comment]
Depends on new column added in https://github.com/beyondessential/tupaia/pull/1291, marking as WIP until then but can be reviewed

File: packages/database/src/migrations/20200910033340-InsertLaosCitiesForWeatherMap-modifies-data.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    const point = { type: 'Point', coordinates: [city.lon, city.lat] };
+
+    rowSql +=
+      `'${city.id}', ` +
+      `'${city.code}', ` +
+      `'${parentId}', ` +
+      `'${city.name}', ` +
+      `'city', ` +
+      `ST_Force2D(ST_GeomFromGeoJSON('${JSON.stringify(point)}')), ` +
+      `'LA', ` +
```
Comment (edmofro) [inline packages/database/src/migrations/20200910033340-InsertLaosCitiesForWeatherMap-modifies-data.js:145]: Should add bounds here as well, for consistency (all entities with a `point` or `region` set currently have `bounds`)

**Review by edmofro — CHANGES_REQUESTED** [review]
Nice code! Just one point inline, but I also have a more fundamental question that I'm struggling to resolve: should `city` be an additional entity type. There are actually already some cities in tupaia, but it's generally because there's a district of the same name. 

One option would be to find an entity that aligns with the city boundarie to add a `point` to, representing the city centre, if these exist in Laos (for example, does `Vientiane Capital` represent the boundaries of Vientiane the city?)

Could you investigate that a little before we commit to adding cities?

A follow up question, if we do add these cities, is whether we want them to exist a little further down the entity hierarchy. I imagine the Laos EOC project will want to go dive pretty deep, and at least have cities sit within provinces of some sort - possibly a question for the product owner.

**Comment by IgorNadj** [comment]
> Could you investigate that a little before we commit to adding cities?

@edmofro  Yep, it would be a change from 10 most populous cities to the 18 province capitals. It doesn't map directly but I think PO/client will be fine with this, I chose it being the 10 most populous cities but province capitals makes more sense. I will make this change.

https://www.wikiwand.com/en/Administrative_divisions_of_Laos

**Comment by IgorNadj** [comment]
> Nice code! Just one point inline, but I also have a more fundamental question that I'm struggling to resolve: should `city` be an additional entity type. There are actually already some cities in tupaia, but it's generally because there's a district of the same name.
> 
> One option would be to find an entity that aligns with the city boundarie to add a `point` to, representing the city centre, if these exist in Laos (for example, does `Vientiane Capital` represent the boundaries of Vientiane the city?)
> 
> Could you investigate that a little before we commit to adding cities?
> 
> A follow up question, if we do add these cities, is whether we want them to exist a little further down the entity hierarchy. I imagine the Laos EOC project will want to go dive pretty deep, and at least have cities sit within provinces of some sort - possibly a question for the product owner.

Discussed. We will go with city.

We will leave nesting under districts for the future if needed (at this point will likely need to hide district labels which display up on top of markers which is not ideal).

File: packages/database/src/migrations/20200910033340-InsertLaosCitiesForWeatherMap-modifies-data.js (hunk 1 above)
Comment (IgorNadj) [reply packages/database/src/migrations/20200910033340-InsertLaosCitiesForWeatherMap-modifies-data.js]: Added

**Review by edmofro — APPROVED** [review]
Looks good to me!

Might want to ask whether @MeganLane29 wants to check the spelling and locations of the province capitals with the project, but the code is all approved :+1:

**Comment by IgorNadj** [comment]
@MeganLane29 here are the spellings I found online, let me know if anything needs to be updated.

```
const PROVINCES = {
  Attapeu: {
    capital: {
      name: 'Attapeu',
      lat: 14.82,
      lon: 106.820556,
    },
  },
  Bokeo: {
    capital: {
      name: 'Houayxay',
      lat: 20.263056,
      lon: 100.433611,
    },
  },
  Bolikhamsai: {
    capital: {
      name: 'Pakxan',
      lat: 18.396389,
      lon: 103.655833,
    },
  },
  Champasak: {
    capital: {
      name: 'Pakse',
      lat: 15.116667,
      lon: 105.783333,
    },
  },
  Houaphanh: {
    capital: {
      name: 'Xam Neua',
      lat: 20.415,
      lon: 104.048,
    },
  },
  Khammouane: {
    capital: {
      name: 'Thakhek',
      lat: 17.4,
      lon: 104.8,
    },
  },
  'Luang Namtha': {
    capital: {
      name: 'Luang Namtha',
      lat: 20.95,
      lon: 101.4,
    },
  },
  'Luang Prabang': {
    capital: {
      name: 'Luang Prabang',
      lat: 19.883333,
      lon: 102.133333,
    },
  },
  Oudomxay: {
    capital: {
      name: 'Muang Xay',
      lat: 20.691389,
      lon: 101.986111,
    },
  },
  Phongsaly: {
    capital: {
      name: 'Phongsali',
      lat: 21.683333,
      lon: 102.1,
    },
  },
  Sainyabuli: {
    capital: {
      name: 'Sainyabuli',
      lat: 19.416667,
      lon: 103.166667,
    },
  },
  Salavan: {
    capital: {
      name: 'Salavan',
      lat: 15.716667,
      lon: 106.416667,
    },
  },
  Savannakhet: {
    capital: {
      name: 'Savannakhet',
      lat: 16.55,
      lon: 104.75,
    },
  },
  Sekong: {
    capital: {
      name: 'Sekong',
      lat: 15.344444,
      lon: 106.7175,
    },
  },
  'Vientiane Prefecture': {
    capital: {
      name: 'Vientiane',
      lat: 17.966667,
      lon: 102.6,
    },
  },
  'Vientiane Province': {
    capital: {
      name: 'Muang Phôn-Hông',
      lat: 18.5,
      lon: 102.416667,
    },
  },
  Xiangkhouang: {
    capital: {
      name: 'Phonsavan',
      lat: 19.416667,
      lon: 103.166667,
    },
  },
  Xaisomboun: {
    capital: {
      name: 'Anouvong',
      lat: 18.9057,
      lon: 103.092,
    },
  },
};
```

**Comment by MeganLane29** [comment]
Thanks @IgorNadj I will confirm these with PSI and let you know if any changes required.

**Comment by MeganLane29** [comment]
@IgorNadj I have received feedback today and there will need to be some updates to the spelling and capital city names. Please see document attached. 

I am still in the process of confirming which Vientiane Capital cities are required to be displayed and will get back to you on this. 

[Laos Province Capitals and Coordinates.docx](https://github.com/beyondessential/tupaia/files/5345329/Laos.Province.Capitals.and.Coordinates.docx)

**Comment by IgorNadj** [comment]
@MeganLane29 ah nice, thanks for that! Yeah looks like it mostly lines up, just the multiple capitals in Vientiane, I think that would be too cluttered on the map. Let me know how you want to proceed. Thanks!

**Comment by MeganLane29** [comment]
@IgorNadj Yep only a couple changes which is good. 

Agree re being a bit cluttered and keen to understand the value of showing weather for multiple cities there versus in other provinces. Will get back to you asap.

**Comment by MeganLane29** [comment]
@IgorNadj Please see updated locations list. Client has confirmed to display only Sisattanak for Vientiane Capital. Thanks!

[Laos Province Capitals and Coordinates.docx](https://github.com/beyondessential/tupaia/files/5345676/Laos.Province.Capitals.and.Coordinates.docx)

**Pull request merged**